pjsip/src/pjsua-lib/pjsua_call.c

```diff
@@ -2811,16 +2778,175 @@ PJ_DEF(pj_status_t) pjsua_call_hangup(pjsua_call_id call_id,
+ * call state.
+ */
+PJ_DEF(pj_status_t) pjsua_call_hangup(pjsua_call_id call_id,
+				      unsigned code,
+				      const pj_str_t *reason,
+				      const pjsua_msg_data *msg_data)
+{
+    pjsua_call *call;
+    pjsip_dialog *dlg = NULL;
+    pj_status_t status;
+
+    if (call_id<0 || call_id>=(int)pjsua_var.ua_cfg.max_calls) {
+	PJ_LOG(1,(THIS_FILE, "pjsua_call_hangup(): invalid call id %d",
+			     call_id));
+    }
+
+    PJ_ASSERT_RETURN(call_id>=0 && call_id<(int)pjsua_var.ua_cfg.max_calls,
+		     PJ_EINVAL);
+
+    PJ_LOG(4,(THIS_FILE, "Call %d hanging up: code=%d..", call_id, code));
+    pj_log_push_indent();
+
+    status = acquire_call("pjsua_call_hangup()", call_id, &call, &dlg);
+    if (status != PJ_SUCCESS)
+	goto on_return;
+
+    if (!call->hanging_up) {
+	pjsip_event user_event;
+
+	pj_gettimeofday(&call->dis_time);
+	if (call->res_time.sec == 0)
+	    pj_gettimeofday(&call->res_time);
+
+    	if (code==0) {
+	    if (call->inv && call->inv->state == PJSIP_INV_STATE_CONFIRMED)
+	        code = PJSIP_SC_OK;
+	    else if (call->inv && call->inv->role == PJSIP_ROLE_UAS)
+	    	code = PJSIP_SC_DECLINE;
+	    else
+	    	code = PJSIP_SC_REQUEST_TERMINATED;
+    	}
+    	
+	call->last_code = code;
+	pj_strncpy(&call->last_text,
+	    	   pjsip_get_status_text(call->last_code),
+		   sizeof(call->last_text_buf_));
+
+    	/* Stop reinvite timer, if it is active. */
+    	if (call->reinv_timer.id) {
+	    pjsua_cancel_timer(&call->reinv_timer);
+	    call->reinv_timer.id = PJ_FALSE;
+    	}
+
+    	/* If media transport creation is not yet completed, we will continue
+    	 * from the media transport creation callback instead.
+         */
+    	if ((call->med_ch_cb && !call->inv) ||
+	    ((call->inv != NULL) &&
+	     (call->inv->state == PJSIP_INV_STATE_NULL)))
+    	{
+            PJ_LOG(4,(THIS_FILE, "Will continue call %d hangup upon "
+                             	 "completion of media transport", call_id));
+
+	    if (call->inv && call->inv->role == PJSIP_ROLE_UAS)
+	    	call->async_call.call_var.inc_call.hangup = PJ_TRUE;
+	    else
+	    	call->async_call.call_var.out_call.hangup = PJ_TRUE;
+
+            if (reason) {
+            	pj_strncpy(&call->last_text, reason,
+		       	   sizeof(call->last_text_buf_));
+            }
+
+	    call->hanging_up = PJ_TRUE;
+    	} else {
+    	    /* Destroy media session. */
+    	    pjsua_media_channel_deinit(call_id);
+
+	    call->hanging_up = PJ_TRUE;
+	    pjsua_check_snd_dev_idle();
+	}
+
+    	/* Call callback which will report DISCONNECTED state.
+    	 * Use user event rather than NULL to avoid crash in
+	 * unsuspecting app.
+	 */
+	PJSIP_EVENT_INIT_USER(user_event, 0, 0, 0, 0);
+    	if (pjsua_var.ua_cfg.cb.on_call_state) {
+	    (*pjsua_var.ua_cfg.cb.on_call_state)(call->index,
+	    					 &user_event);
+	}
+
+    }
+
+    if (call->inv)
+    	call_inv_end_session(call, code, reason, msg_data);
```

Comment:
If `pjsua_call_hangup()` is called multiple times (for the same call), is it handled already (e.g: avoid rescheduling an active timer)? The scenario is perhaps not uncommon, e.g: hangup button in the UI not disabled immediately so the user may press it multiple times (by accident or not) and the app code simply invokes `pjsua_call_hangup()` whenever the button is pressed.

Reply:
Interesting scenario.

Here's the result of my test:
When `pjsua_call_hangup() is called more than once, each will send a 'BYE' request. This is consistent with the current behaviour.
But with this patch, the scenario (trying to hangup call more than once) can be easily avoidable since app will get the DISCONNECTED state immediately during the first hangup attempt.

I have added the timer check for more robustness.